trimer guess

jake163 opened the pull request.

Commit:
trimer guess

jake163 commented on cluster_density_overlap.py:
```diff
@@ -0,0 +1,86 @@
+
+bs = "aug-cc-pvdz"
+rho = density(mol, bs)
+nitr = run_scf(mol, bs, rho)
```
Right here is where the error I mentioned comes up.

jake163 commented on cluster_density_overlap.py:
```diff
@@ -0,0 +1,86 @@
+overlap = overlap_dens(41, 82)
+dimer_dens_diag = block_diag(density_1, density_1)
+rho = np.subtract(dimer_dens_diag, overlap)
+nitr = run_scf(mol, bs, rho)
```
Right here is where the error comes up

jake163 replied:
My previous comment was on the wrong line.

ryanmrichard reviewed (commented):
Try doing three hydrogen dimers in STO-3G basis. 

Let the dimers be A, B, and C such that hydrogens 1 and 2 are in A, 3 and 4 are in B, and 5 and 6 are in C. Since it's STO-3G each hydrogen gets a single 1s orbital, which we just identify by the number of the hydrogen (e.g., basis function 2 will be the 1s orbital on hydrogen 2).

We need to compute six density matrices. The first three are the 2 by 2 densities of each dimer:
- P^A (the density of dimer A) rows and columns are basis functions 1 and 2
- P^B (the density of dimer B) rows and columns are basis functions 3 and 4
- P^C (the density of dimer C) rows and columns are basis functions 5 and 6
The next three are the 4 by 4 densities of each pair of dimers:
- P^{AB} (the density of the pair AB) rows and columns are basis functions 1, 2, 3, and 4
- P^{AC} (the density of the pair AC) rows and columns are basis functions 1, 2, 5, and 6
- P^{BC} (the density of the pair BC) rows and columns are basis functions 3, 4, 5, and 6

Next we need to project these six matrices into the trimer basis set. A matrix in the trimer basis set is a 6 by 6 matrix whose rows and columns correspond to basis functions 1, 2, 3, 4, 5, 6 by 1, 2, 3, 4, 5, 6. In the trimer basis set P^A looks like (P^A_{ij} is the i,j-th element of P^A where i,j are element indices **not** basis indices, i.e. P^B_{11} is the element between 1s orbitals of hydrogen 3 not hydrogen 1):

```
P^A_{11} P^A_{12} 0 0 0 0
P^A_{21} P^A_{22} 0 0 0 0
    0        0    0 0 0 0
    0        0    0 0 0 0
    0        0    0 0 0 0
    0        0    0 0 0 0
```
P^B looks like:

```
0 0     0        0    0 0
0 0     0        0    0 0
0 0 P^B_{11} P^B_{12} 0 0
0 0 P^B_{21} P^B_{22} 0 0
0 0     0        0    0 0
0 0     0        0    0 0
```

P^C looks like:

```
0 0 0 0     0        0
0 0 0 0     0        0
0 0 0 0     0        0
0 0 0 0     0        0
0 0 0 0 P^C_{11} P^C_{12}
0 0 0 0 P^C_{21} P^C_{22}
```

P^{AB} looks like:
```
P^{AB}_{11} P^{AB}_{12} P^{AB}_{13} P^{AB}_{14} 0 0
P^{AB}_{21} P^{AB}_{22} P^{AB}_{23} P^{AB}_{24} 0 0
P^{AB}_{31} P^{AB}_{32} P^{AB}_{33} P^{AB}_{34} 0 0
P^{AB}_{41} P^{AB}_{42} P^{AB}_{43} P^{AB}_{44} 0 0
      0           0           0           0     0 0
      0           0           0           0     0 0
```

P^{AC} looks like:
```
P^{AC}_{11} P^{AB}_{12} 0 0 P^{AC}_{13} P^{AC}_{14} 
P^{AC}_{21} P^{AB}_{22} 0 0 P^{AC}_{23} P^{AC}_{24}
      0           0     0 0       0           0 
      0           0     0 0       0           0
P^{AC}_{31} P^{AC}_{32} 0 0 P^{AC}_{33} P^{AC}_{34}
P^{AC}_{41} P^{AC}_{42} 0 0 P^{AC}_{43} P^{AC}_{44}
```

and P^{BC} looks like:
```
0 0       0           0           0           0
0 0       0           0           0           0   
0 0 P^{BC}_{11} P^{BC}_{12} P^{BC}_{13} P^{BC}_{14}
0 0 P^{BC}_{21} P^{BC}_{22} P^{BC}_{23} P^{BC}_{24}
0 0 P^{BC}_{31} P^{BC}_{32} P^{BC}_{33} P^{BC}_{34}
0 0 P^{BC}_{41} P^{BC}_{42} P^{BC}_{43} P^{BC}_{44}
```
Your guess is then P^{AB} + P^{AC} + P^{BC} - P^A - P^B - P^C (all of these matrices in the trimer basis)

jake163 commented:
There were a lot of interactions that I wasn't accounting for.

jake163 commented:
I'd like to ask a conceptual question if you don't mind (the learning-undergraduate variety). 

If you had a wavefunction that was a function of three spatial coordinates, square that for little p, then for a finite basis set **P** would be a rank three tensor array, no? Would that account for more interaction?

ryanmrichard commented:
In the context of SCF we are usually talking about the reduced one-electron density in the AO basis set. This is always a rank 2 tensor (it's number of AOs by number of AOs). The size of the tensor only depends on the number of AOs; turning on or off interactions has no effect on the size, but changes the values of the elements. 

The analysis in terms of the wavefunction is much more complicated because in SCF it is a Slater determinant formed from the MOs, which themselves are LCAOs. We usually talk about matrix elements of the wavefunction as then we get to integrate over the degrees of freedom and things are simpler. Off the top of my head I'm not sure what the wavefunction squared actually looks like for SCF.

jake163 committed:
Update cluster_density_overlap.py

jake163 commented on cluster_density_overlap.py:
```diff
@@ -0,0 +1,73 @@
+c = np.subtract(b, P_1)
+d = np.subtract(c, P_2)
+rho = np.subtract(d, P_3)
+nitr = run_scf(mol, bs, rho)
```
I get 7 outputted energies, with the final one being with our guess (5 cycles).

jake163 replied:
I get 7 outputted energies, with the final one being with our guess (5 cycles).

jake163 committed:
Update cluster_density_overlap.py

jake163 commented on cluster_density_overlap.py:
```diff
@@ -0,0 +1,71 @@
+density_6 = density(H4C, bs, verbose = 5, unpaired_electrons = 0)
+P_6 = block_diag(zero_m2, density_6)
+
+P_Di = [P_4, P_5, P_6]
```
I made the matrix addition/subtraction steps a bit more streamlined.

jake163 committed:
Update cluster_density_overlap.py

This is what I got from Jonathan; patch together a guess density from dimer densities.

jake163 committed:
Rename cluster_density_overlap.py to Many-body-guess

ryanmrichard commented on Many-body-guess:
```diff
@@ -0,0 +1,31 @@
+A[4:6, 2:6] = h4BC[2:4, 0:4]
+A[2:4, 4:6] = h4BC[0:2, 2:4]
+A[4:6, 0:2] = h4AC[2:4, 0:2]
+A[0:2, 4:6] = h4AC[0:2, 2:4]
```
This doesn't look right to me. In particular you're overwriting AA blocks, and not adding to them, for example. Also you're not subtracting out the monomer densities.

jake163 replied:
The final density is not what I put in the slack chat yesterday? My misunderstanding. I thought we had to subtract out an AA because you get two of those subblocks, from A-B density and A-C density. So i used full A-B, but only parts of B-C and A-C to avoid the overcounting of the monomer subblocks AA, BB, and CC. 

I thought this is what Jonathan was talking about.

jake163 replied:
And so I thought that since there was no overcounting with this piecewise construction of the guess rho, no addition or subtraction.

jake163 replied:
Where did my thinking go wrong here, if you don't mind saying?

jwaldrop107 replied:
The way this is currently written loses some of the desired interactions. The interaction between A and B is mostly contained in the AB and BA blocks of the A-B density, but there are effects on the AA and BB parts as well. So the AA block of the guess density needs to be AA_{A-B} + AA_{A-C} - AA_{A}, so that all desired interactions are captured in the guess.

You can use straight assignment for the off-diagonal blocks, but the arithmetic is required (in some fashion) for the diagonal blocks.

jake163 committed:
Rename Many-body-guess to Many-body-guess.py

jake163 commented on Many-body-guess.py:
```diff
@@ -0,0 +1,31 @@
+from wrap_pyscf import *
```
I added a .py extension

jake163 commented:
Ok, thanks for explaining. I will fix it.

On Thu, Aug 20, 2020 at 3:55 PM Jonathan M. Waldrop <
notifications@github.com> wrote:

> *@jwaldrop107* commented on this pull request.
> ------------------------------
>
> In Many-body-guess
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#discussion_r474267084>
> :
>
> > +
> +h2A = density(mol[0:2], bs)
> +h2B = density(mol[2:4], bs)
> +h2C = density(mol[4:6], bs)
> +h4AB = density(mol[0:4], bs)
> +h4AC = density((mol[0:2]+mol[4:6]), bs)
> +h4BC = density(mol[2:6], bs)
> +#density_list = [h2A, h2B, h2C, h4AB, h4AC, h4BC]
> +
> +A = np.zeros((6,6))
> +
> +A[0:4, 0:4] = h4AB
> +A[4:6, 2:6] = h4BC[2:4, 0:4]
> +A[2:4, 4:6] = h4BC[0:2, 2:4]
> +A[4:6, 0:2] = h4AC[2:4, 0:2]
> +A[0:2, 4:6] = h4AC[0:2, 2:4]
>
> The way this is currently written loses some of the desired interactions.
> The interaction between A and B is mostly contained in the AB and BA blocks
> of the A-B density, but there are effects on the AA and BB parts as well.
> So the AA block of the guess density needs to be AA_{A-B} + AA_{A-C} -
> AA_{A}, so that all desired interactions are captured in the guess.
>
> You can use straight assignment for the off-diagonal blocks, but the
> arithmetic is required (in some fashion) for the diagonal blocks.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#discussion_r474267084>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AP37KM6OBVJ4PTIFX2I4O33SBWEV3ANCNFSM4PWZZS2A>
> .
>

jake163 committed:
Update Many-body-guess.py

jake163 commented on Many-body-guess.py:
```diff
@@ -0,0 +1,35 @@
+
+rho = A
+nitr = run_scf(mol, bs, rho)
+print(nitr)
```
4 cycles

ryanmrichard replied:
How many does (H2)3 take with PySCF's default guess?

jake163 commented:
Also four.

On Fri, Aug 21, 2020, 9:18 AM Ryan Richard <notifications@github.com> wrote:

> *@ryanmrichard* commented on this pull request.
> ------------------------------
>
> In Many-body-guess.py
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#discussion_r474727602>
> :
>
> > +
> +A = np.zeros((6,6))
> +
> +A[0:2, 0:2] = h4AB[0:2, 0:2]+h4AC[0:2, 0:2]-h2A
> +A[2:4, 2:4] = h4AB[2:4, 2:4]+h4BC[0:2, 0:2]-h2B
> +A[4:6, 4:6] = h4AC[2:4, 2:4]+h4BC[2:4, 2:4]-h2C
> +A[2:4, 0:2] = h4AB[2:4, 0:2]
> +A[0:2, 2:4] = h4AB[0:2, 2:4]
> +A[2:4, 4:6] = h4BC[0:2, 2:4]
> +A[4:6, 2:4] = h4BC[2:4, 0:2]
> +A[4:6, 0:2] = h4AC[2:4, 0:2]
> +A[0:2, 4:6] = h4AC[0:2, 2:4]
> +
> +rho = A
> +nitr = run_scf(mol, bs, rho)
> +print(nitr)
>
> How many does (H2)3 take with PySCF's default guess?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#discussion_r474727602>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AP37KM35KXNA7QULRUNOKZLSBZ63VANCNFSM4PWZZS2A>
> .
>

jwaldrop107 commented:
I looked at the output for the two. It's the same number of cycles, but the new guess is considerably better converged.

ryanmrichard commented:
hydrogen molecules aren't necessarily the strongest interacting things in the world, I'd be curious to see how the situation changes with water molecules where polarization matters a lot more.

jake163 commented:
Which basis set?

On Fri, Aug 21, 2020 at 12:18 PM Ryan Richard <notifications@github.com>
wrote:

> hydrogen molecules aren't necessarily the strongest interacting things in
> the world, I'd be curious to see how the situation changes with water
> molecules where polarization matters a lot more.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#issuecomment-678399687>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AP37KMYOIVNPKFZ5OT4NMNLSB2T7TANCNFSM4PWZZS2A>
> .
>

ryanmrichard commented:
I'd do aug-cc-pvdz as the basis set for water. Note that you're going to have to generalize your script more as it's no longer going to be 2 basis functions per monomer.

jake163 commented:
True. Ok, I'll see about the trimer.

On Fri, Aug 21, 2020 at 1:24 PM Ryan Richard <notifications@github.com>
wrote:

> I'd do aug-cc-pvdz as the basis set for water. Note that you're going to
> have to generalize your script more as it's no longer going to be 2 basis
> functions per monomer.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#issuecomment-678426704>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AP37KM6RENRH3M4NWQ4GZ5DSB23UTANCNFSM4PWZZS2A>
> .
>

jake163 committed:
Update Many-body-guess.py

jake163 commented on Many-body-guess.py:
```diff
@@ -0,0 +1,79 @@
+
+rho = A
+nitr = run_scf(mol, bs, rho)
+print(nitr)
```
This guess: 6 cycles
Standard: 9 cycles

ryanmrichard replied:
Alright, we're getting somewhere!!! Try doing water tetramer. The full density looks like:

```
AA AB AC AD
BA BB BC BD
CA CB CC CD
DA DB DC DD
```
Now however, AA is going to get contributions from dimers AB, AC, and AD which overcounts A twice.

In general for a system with N monomers, the I-th diagonal block is going to be given by:
![image](https://user-images.githubusercontent.com/5280139/90962309-bf987380-e474-11ea-95f6-5f25b1c268e7.png)
where IJ is the II block of dimer IJ's density and I (on the right) is the monomer density. The off-diagonal blocks are going to continue to just be the contribution from the respective dimer, *i.e.* the IJ and JI blocks will only have contributions from dimer IJ).

jake163 replied:
Yeah, I cheered a bit when I got the results. Startled my roommate 😆 

I can get working on it. For this next bit of code, do you want me to paste under the trimer, paste over the other scripts, or begin a new PR?

ryanmrichard commented:
I'd go below it.

On Sat, Aug 22, 2020, 2:20 PM jake163 <notifications@github.com> wrote:

> *@jake163* commented on this pull request.
> ------------------------------
>
> In Many-body-guess.py
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#discussion_r475122055>
> :
>
> > +A = np.zeros((M*3,M*3))
>
> +
>
> +#Monomer_density = block_diag(H2OA, H2OB, H2OC)
>
> +A[0:M, 0:M] = H4O2AB[0:M, 0:M]+H4O2AC[0:M, 0:M]-H2OA
>
> +A[M:N, M:N] = H4O2AB[M:N, M:N]+H4O2BC[0:M, 0:M]-H2OB
>
> +A[N:M*3, N:M*3] = H4O2AC[M:N, M:N]+H4O2BC[M:N, M:N]-H2OC
>
> +A[0:M, M:N] = H4O2AB[0:M, M:N]
>
> +A[M:N, 0:M] = H4O2AB[M:N, 0:M]
>
> +A[0:M, N:M*3] = H4O2AC[0:M, M:N]
>
> +A[N:M*3, 0:M] = H4O2AC[M:N, 0:M]
>
> +A[M:N, N:M*3] = H4O2BC[0:M, M:N]
>
> +A[N:M*3, M:N] = H4O2BC[M:N, 0:M]
>
> +
>
> +rho = A
>
> +nitr = run_scf(mol, bs, rho)
>
> +print(nitr)
>
>
> Yeah, I cheered a bit when I got the results. Startled my roommate 😆
>
> I can get working on it. For this next bit of code, do you want me to
> paste under the trimer, paste over the other scripts, or begin a new PR?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rmrresearch/SCF-Guess-Project/pull/12#discussion_r475122055>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIJDC25H37PI6SCYBTT5CDSCAK6ZANCNFSM4PWZZS2A>
> .
>

jake163 committed:
Update Many-body-guess.py

jake163 commented on Many-body-guess.py:
```diff
@@ -0,0 +1,133 @@
+
+rho = A
+nitr = run_scf(mol, bs, rho)
+print(nitr)
```
Unfortunately I got 9 cycles with this script, which is the same as the atomic guess. There could be an error in the script, but I haven't found one. 

If considering dimer interactions improved the trimer guess, would trimer interactions improve the guess for the tetramer?

jwaldrop107 replied:
Reasonable that the trimers would help the tetramer guess. You could kind of stack the process. Use monomers to guess dimers, dimers for trimers, and so on.

ryanmrichard replied:
If we have to go to trimers to get a good guess this idea is dead in the water (no pun intended). The MBE is pretty converged by the three-body term for most systems (water's significant polarization makes it an exception; for water it converges with four-body terms). In other words, if we have to do trimers the work required to get the guess is the same amount of work required to compute the energy of the full system (for most systems)...

ryanmrichard replied:
Next I'd try writing the more general two-body algorithm and applying it to the water pentamer and hexamer.

jwaldrop107 commented on Many-body-guess.py:
```diff
@@ -0,0 +1,133 @@
+
+
+from wrap_pyscf import *
+import numpy as np
```
You only have to have these imports one time. No need to keep importing them.
